### PR TITLE
Add usage of select  instead of pulling for Linux based platform (PCAN Interface)

### PR DESCRIPTION
**Pull request by BKaDamien**
Avoid pulling of "time.sleep(0.001)" in pcan interface method _recv_internal, and use select python function instead.

### CODE REVIEW

**Comment by BKaDamien** [comment]
Hi python-can maintainers/developers,

I need help regarding my changes related to the introduction of select for linux platform:

The issue come from test_recv_no_message from test module test_pcan, when I run the sae code everything is fine the test_recv_no_message is passed. 

below my setup:
test run on : Ubuntu 20.04.3 LTS
python :  3.8.10

But since the beginning of this PR, the Ubuntu tests  are constantly failing at test_recv_no_message. Based on the trace from pytest it seems that despite of the select mock the return is not executed  and the code try to access an attribute from a None Value(see link below).
https://github.com/hardbyte/python-can/actions/runs/3273689147/jobs/5402946921

Thanks for your support

**Comment by felixdivo** [comment]
Looks good! I haven't tested it with hardware though. We should probably do that.

**Review by felixdivo — APPROVED** [review]

**Comment by BKaDamien** [comment]
> Looks good! I haven't tested it with hardware though. We should probably do that.

Tested on my side, this update will be a little less consumer than the polling.

**Comment by felixdivo** [comment]
Sorry for being this pedantic, but on which platforms did you try this? There appear to be 4 variants here: Windows and Linux and each with and without polling. For each of the 4, sending and receiving has to work. :slightly_smiling_face:

**Comment by BKaDamien** [comment]
> Sorry for being this pedantic, but on which platforms did you try this? There appear to be 4 variants here: Windows and Linux and each with and without polling. For each of the 4, sending and receiving has to work. 🙂

Based on my changes:
Tested on Windows -> OK (use WaitForSingleObject windows base event)
Test on Ubuntu -> OK (use python select base event)
For the other MAC, .... -> Just use the polling

**Comment by felixdivo** [comment]
Okay, thank you for documenting that here. 🙂 I think this is sufficient, since the polling implementation id not really change and there are also unit tests that make sure nothing gets too broken.

I'd propose to merge it as it.

**Comment by BKaDamien** [comment]
> Okay, thank you for documenting that here. 🙂 I think this is sufficient, since the polling implementation id not really change and there are also unit tests that make sure nothing gets too broken.
> 
> I'd propose to merge it as it.

You are the boss :)

**Comment by BKaDamien** [comment]
Arf I wrongly click and close I am sorry

**Comment by felixdivo** [comment]
@hardbyte You can merge it before or after the 4.1.0 release. If you add it before, we'd have to amend the changelog.

File: can/interfaces/pcan/pcan.py [hunk 1]
```diff
@@ -437,11 +456,26 @@ def set_device_number(self, device_number):
...
+
     def _recv_internal(self, timeout):
 
-        if HAS_EVENTS:
+        if HAS_EVENTS and IS_WINDOWS:
             # We will utilize events for the timeout handling
             timeout_ms = int(timeout * 1000) if timeout is not None else INFINITE
+        elif HAS_EVENTS and IS_LINUX:
+            # We will utilize events for the timeout handling
+            timeout_s = timeout if timeout != None else None
```
Comment (zariiii9003) [inline can/interfaces/pcan/pcan.py:478]: This seems redundant

File: can/interfaces/pcan/pcan.py [hunk 2]
```diff
@@ -64,7 +64,8 @@
 log = logging.getLogger("can.pcan")
 
 MIN_PCAN_API_VERSION = version.parse("4.2.0")
-
+IS_WINDOWS = PLATFORM == "Windows"
+IS_LINUX = PLATFORM == "Linux"
```
Comment (felixdivo) [inline can/interfaces/pcan/pcan.py:68]: We could also move these to `can/interfaces/pcan/basic.py` and just import `IS_WINDOWS` & `IS_LINUX`, and also change the if around `import winreg` in there to then use `if IS_WINDOWS` instead. Small thing tough.

File: can/interfaces/pcan/pcan.py (hunk 1 above)
Comment (BKaDamien) [reply can/interfaces/pcan/pcan.py:477]: I don't think so select accept timeout in seconds or None.
On windows WaitForSingleObject expects a timeout in ms

**Comment by zariiii9003** [comment]
@BKaDamien I cleaned up a bit, please take a look.

**Comment by BKaDamien** [comment]
> @BKaDamien I cleaned up a bit, please take a look.

It's fine. Thanks for the chnages.

**Review by zariiii9003 — APPROVED** [review]

**Pull request merged**